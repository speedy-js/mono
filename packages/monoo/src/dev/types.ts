/**
 * Module dependencies
 */
import { IMonorepoBuilderOptions } from '@monoo/types';

export interface IDOptions extends IMonorepoBuilderOptions {}
