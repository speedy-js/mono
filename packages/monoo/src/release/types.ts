/**
 * Module dependencies
 */
import type { RequestInit } from "node-fetch";
import type urlJoin from "url-join";
import { IPackageJSON } from "@monoo/types"

/**
 * Shared namespace
 */
export namespace SharedNS {
  export interface IOptions {
    cwd?: string;
  }
}

/**
 * Release namespace
 */
export namespace ReleaseNS {
  export interface IModifyReleaseArgumentsContext {
    arguments: string[];
    version: string;
    tag: string;
  }

  export interface IReadyContext {
    cwd: string;
  }

  export interface IReleasedContext {
    cwd: string;
    version: string;
    tag: string;
  }

  export interface IPromptAnswers {
    bump: string;
    customVersion: string;
    npmTag: string;
    customNpmTag: string;
  }

  export interface IOptions extends SharedNS.IOptions {
    /**
     * Generate changelog or not.
     */
    changelog?: boolean;
    /**
     * Run in band or not.
     */
    runInBand?: boolean;
    /**
     * Ignore scripts, it will ignore the npm life cycle hooks (e.g. prepublishOnly).
     */
    ignoreScripts?: boolean;
    /**
     * Execute custom build script before release.
     *
     * It is recommended to run topological build before actual publishing process,
     * and then ignore the build script (with `ignoreScripts`) of each package during
     * actual publishing release.
     */
    build?: boolean | string;
    /**
     * Preview execution.
     */
    dryRun?: boolean;
    /**
     * @deprecated
     */
    ready?: (ctx: IReadyContext) => Promise<unknown>;
    /**
     * @deprecated
     */
    modifyReleaseArguments?: (ctx: IModifyReleaseArgumentsContext) => string[];
    /**
     * @deprecated
     */
    released?: (ctx: IReleasedContext) => Promise<unknown>;
  }
}

/**
 * Run namespace
 */
export namespace RunNS {
  export interface IOptions extends SharedNS.IOptions {
    commands: string[];
    runInBand: boolean;
    filter?(v: IPackageJSON): boolean;
    [key: string]: unknown;
  }
}

/**
 * Disttag namespace
 */
export namespace DistTagNS {
  export interface IOptions extends SharedNS.IOptions {
    distVersion: string;
    tag: string;
  }
}

/**
 * Path namespace
 */
export namespace PatchNS {
  export interface IOptions extends SharedNS.IOptions {
    version: string;
    tag: string;
    runInBand: boolean;
    ignoreScripts: boolean;
  }
}

/**
 * Dingding namespace
 */
export namespace DingdingNS {
  export interface IWebhook {
    name: string;
    webhook: string;
    secret: string;
  }

  export interface IOptions extends IWebhook {
    body?: {
      msgtype: "markdown";
      markdown: {
        title: string;
        text: string;
      };
    };
    example?: boolean;
  }

  export interface IWebhookConfig {
    /**
     * Webhooks to post
     */
    webhooks: DingdingNS.IWebhook[];
    /**
     * Test webhook
     */
    testWebhook?: DingdingNS.IWebhook;
    /**
     * Using test webhook or not
     */
    test?: boolean;
  }
}

/**
 * Yuque namespace
 */
export namespace YuqueNS {
  /**
   * Transformer context
   */
  export interface ITransformerContext {
    host: string;
    urlJoin: typeof urlJoin;
    transformerOptions: Record<string, unknown>;
    options: IOptions;
  }

  /**
   * Post data
   */
  export interface IYuquePostData {
    title: string;
    cover: string;
    description: string;
    custom_description: string;
    book: {
      namespace: string;
      user: {
        avatar_url: string;
      };
    };
    slug: string;
  }

  /**
   * Transformer
   */
  export type Transformer = (
    data: IYuquePostData,
    ctx: ITransformerContext
  ) => Record<string, unknown>;

  /**
   * Options
   */
  export interface IOptions
    extends SharedNS.IOptions,
      DingdingNS.IWebhookConfig {
    /**
     * Url of yuque post
     */
    url?: string;
    /**
     * Host of yuque
     * @default 'https://yuque.antfin-inc.com/api/v2'
     */
    host?: string;
    /**
     * Cover to post
     */
    cover?: string;
    /**
     * Endpoint of yuque open api
     */
    endpoint?: string;
    /**
     * Request options
     */
    options?: Partial<RequestInit>;
    /**
     * Transformer preset
     */
    transformerPreset?: "dingding-actionCard";
    /**
     * Transformer options
     */
    transformerOptions?: Record<string, unknown>;
    /**
     * Custom transformer
     */
    transformer?: Transformer;
    /**
     * Log response
     */
    log?: boolean;
  }
}

/**
 * changelog namespace
 */
export namespace ChangelogNS {
  export interface ICommitAuthorOptions {
    /**
     * Attach author to each changelog commit
     *
     * @default false
     */
    attachAuthor?: boolean;
    /**
     * Type of author name
     *
     * @default 'name'
     */
    authorNameType?: "name" | "email";
    /**
     * Get author remote page
     *
     * @param author
     */
    getAuthorPage?(author: ICommitAuthor): string;
  }

  export interface IOptions extends SharedNS.IOptions, ICommitAuthorOptions {
    /**
     * current Version
     *
     * @default version from lerna.json
     */
    version?: string;
    /**
     * Beautify changelog, Remove top-level useless header
     *
     * @see https://github.com/conventional-changelog/conventional-changelog/issues/376
     * @default false
     */
    beautify?: boolean;
    /**
     * Create changelog commit
     *
     * @default false
     */
    commit?: boolean;
    /**
     * Create git push action
     *
     * @default false
     */
    gitPush?: boolean;
  }
  /**
   * Commit author
   */
  export interface ICommitAuthor {
    /**
     * Commit author name
     */
    name: string;
    /**
     * Email name
     */
    emailName: string;
    /**
     * Full email
     */
    email: string;
  }
}

/**
 * PostChangelog namespace
 */
export namespace PostChangelogNS {
  /**
   * View all link
   */
  export interface IViewAll {
    text: string;
    url: string;
  }
  /**
   * Options to infer changelog
   */
  export interface IInferChangelogOptions
    extends SharedNS.IOptions,
      ChangelogNS.ICommitAuthorOptions {
    /**
     * Package name
     */
    packageName: string;
    /**
     * changelog file name
     *
     * @default 'CHANGELOG.md'
     */
    file?: string;
    /**
     * changelog preset, avilable presets: conventional-changelog、afx
     *
     * @default 'conventional-changelog'
     */
    preset?: string;
    /**
     * custom changelog parser
     */
    parser?: (
      changelog: string,
      packageName: string
    ) => {
      title: string;
      content: string;
    };
    /**
     * view all config
     */
    viewAll?: IViewAll;
  }

  export interface IOptions
    extends IInferChangelogOptions,
      DingdingNS.IWebhookConfig {
    /**
     * Custom title, defaults to inferred title from changelog.
     */
    title?: string;
    /**
     * Custom content, defaults to inferred content from changelog.
     */
    content?: string;
  }
}

export type IConfig = ReleaseNS.IOptions & {
  verbose?: boolean;
  postChangelog?: PostChangelogNS.IOptions;
  yuque?: YuqueNS.IOptions;
  dingding?: DingdingNS.IOptions;
  patch?: PatchNS.IOptions;
  distTag?: DistTagNS.IOptions;
  run?: RunNS.IOptions;
};

export function defineConfig(config: IConfig): IConfig {
  return config;
}
