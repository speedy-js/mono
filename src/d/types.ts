/**
 * Module dependencies
 */
import { IMonorepoBuilderOptions } from '../types';

export interface IDOptions extends IMonorepoBuilderOptions {}
