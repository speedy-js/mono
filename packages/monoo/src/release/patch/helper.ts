/**
 * Module dependencies
 */
import boxen, { BorderStyle } from "boxen";
import textTable from "text-table";
import stringWidth from "string-width";
import { IMonorepoPackageWithRemoteInfo } from '@monoo/types';
import { chalk } from '@monoo/shared';

/**
 * Get visible release status log
 */
export function getVisibleReleaseStatusLog(
  pkgs: IMonorepoPackageWithRemoteInfo[],
  currentVersion: string,
  currentTag: string
) {
  const table = pkgs.map((pkg) => {
    if (pkg.remoteVersion === currentVersion) {
      return [pkg.name, pkg.version, "Yes", "-"].map((v) => chalk.dim(v));
    }

    return [
      chalk.bold(chalk.red(pkg.name)),
      chalk.bold(chalk.red(pkg.remoteVersion)),
      chalk.bold(chalk.red("NO")),
      chalk.bold(chalk.red(currentVersion)),
    ];
  });

  table.unshift(
    [
      "Package Name",
      `Remote Version (tag: ${currentTag})`,
      "Published",
      "Expected Version",
    ].map((v) => chalk.dim(v))
  );

  return boxen(
    textTable(table, {
      stringLength: stringWidth,
    }),
    {
      borderStyle: BorderStyle.Round,
      padding: 1,
      dimBorder: true,
    }
  )
}
