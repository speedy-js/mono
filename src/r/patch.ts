/**
 * Module dependencies
 */
import { join } from 'path';
import { readFileSync, writeFileSync } from 'fs';
import chalk from 'chalk';
import inquirer from 'inquirer';
import execa from 'execa';
import boxen from 'boxen';
import textTable from 'text-table';
import stringWidth from 'string-width';
import { resolveLernaConfig, resolvePackages, logger } from './shared';
import { PatchNS, NpmNS } from './types';

/**
 * Remove duplicate gitHead field when "lerna publish" fails.
 *
 * @param {String} path
 * @returns {void}
 */

export function removeGitHead(path: string) {
  const GIT_HEAD_REG = /,[\n\t\s]*"gitHead":\s"[^"]+"/;
  let content = readFileSync(path, 'utf-8');
  content = content.replace(GIT_HEAD_REG, '');
  writeFileSync(path, content, 'utf-8');
}

/**
 * Enter a interactive cli to know the current situation of `lerna publish`.
 *
 * @param {String} cwd
 * @param {String} version
 * @param {String} tag
 * @returns {Promise<void>}
 */

export async function patch(options: PatchNS.IOptions): Promise<void | never> {
  const { cwd, tag, runInBand = false, ignoreScripts = false } = options;

  let { version } = options;

  logger.info('Patch Started');

  if (!tag) {
    throw new Error('[MONO] "tag" is required for "patch"');
  }

  if (!version) {
    try {
      const config = resolveLernaConfig(cwd);
      version = config?.data.version;
    } catch (e) {
      // dot not handle it for now
    }
  }

  logger.info(`Version: ${chalk.cyan(version)}`);
  logger.info(`Tag: ${chalk.cyan(tag)}`);

  const pkgs = await resolvePackages({ cwd, tag });
  pkgs.forEach((pkg) => removeGitHead(join(pkg.path, 'package.json')));

  if (pkgs.every((pkg) => pkg.version === version)) {
    return console.log(
      `${chalk.cyan('❯ ')}${chalk.gray(
        '[MONO]',
      )} Do not need "patch" since all packages've been published correctly! `,
    );
  }

  const table = pkgs.map((pkg) => {
    if (pkg.version === version) {
      return [
        chalk.gray(pkg.name),
        chalk.gray(pkg.version),
        chalk.gray('√'),
        '-',
      ];
    }

    return [chalk.cyan(pkg.name), chalk.green(pkg.version), '❌', version];
  });

  table.unshift(
    [
      'Package Name',
      `Remote Version (tag: ${tag})`,
      'Status',
      'Target Version',
    ].map((v) => chalk.dim(v)),
  );

  console.log(
    boxen(
      textTable(table, {
        stringLength: stringWidth,
      }),
    ),
  );

  const { yes } = await inquirer.prompt([
    {
      name: 'yes',
      message: 'Continue to patch',
      type: 'list',
      choices: ['N', 'Y'],
    },
  ]);

  if (yes === 'Y') {
    const patchedPkgs = pkgs.filter((pkg) => pkg.version !== version);

    const releasePkg = async (pkg: NpmNS.IRemotePackageInfo) => {
      const commandCwd = join(cwd, `packages/${pkg.dirname}`);
      console.log(chalk.gray(`$ npm publish --tag ${tag} # ${commandCwd}`));
      const npmArgs = ['publish', '--tag', tag];
      if (ignoreScripts) {
        npmArgs.push('--ignore-scripts');
      }
      await execa('npm', npmArgs, {
        // silent: true,
        stdio: 'inherit',
        cwd: commandCwd,
      });
      console.log(`+ ${pkg.name}@${version}\n`);
    };

    if (runInBand) {
      for (const pkg of patchedPkgs) {
        await releasePkg(pkg);
      }
    } else {
      await Promise.all(patchedPkgs.map(releasePkg));
    }
  }
}
