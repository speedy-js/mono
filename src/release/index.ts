/**
 * Module dependencies
 */
import chalk from "chalk";
import execa from "execa";
import semver, { ReleaseType } from "semver";
import inquirer from "inquirer";
import { patch } from "./patch";
import {
  logger,
  resolveLernaConfig,
  ensureLernaConfig,
  cleanLernaConfig,
} from "./shared";
import { ReleaseNS, ChangelogNS } from "./types";
import { changelog } from "./changelog";

export { patch, changelog };

/**
 * Select version and tag to be released.
 */
async function selectVersionAndTag(currentVersion: string) {
  const customItem = { name: "Custom", value: "custom" };
  const bumps: ReleaseType[] = [
    "patch",
    "minor",
    "major",
    "prerelease",
    "premajor",
  ];
  const versions = bumps.reduce<Record<string, string>>((memo, bump) => {
    memo[bump] = semver.inc(currentVersion, bump);
    return memo;
  }, {});

  const bumpChoices = bumps.map((b) => ({
    name: `${b} (${versions[b]})`,
    value: b,
  }));

  function getVersion(answers: ReleaseNS.IPromptAnswers) {
    return answers.customVersion || versions[answers.bump];
  }

  function getNpmTags(version: string) {
    if (isPreRelease(version)) {
      return ["next", "latest", "beta", customItem];
    }
    return ["latest", "next", "beta", customItem];
  }

  function isPreRelease(version: string) {
    return Boolean(semver.prerelease(version));
  }

  const { bump, customVersion, npmTag, customNpmTag } =
    await inquirer.prompt<ReleaseNS.IPromptAnswers>([
      {
        name: "bump",
        message: "Select release type:",
        type: "list",
        choices: [...bumpChoices, customItem],
      },
      {
        name: "customVersion",
        message: "Input version:",
        type: "input",
        when: (answers) => answers.bump === "custom",
      },
      {
        name: "npmTag",
        message: "Input npm tag:",
        type: "list",
        default: (answers: ReleaseNS.IPromptAnswers) =>
          getNpmTags(getVersion(answers))[0],
        choices: (answers) => getNpmTags(getVersion(answers)),
      },
      {
        name: "customNpmTag",
        message: "Input customized npm tag:",
        type: "input",
        when: (answers) => answers.npmTag === "custom",
      },
    ]);

  const version = customVersion || versions[bump];
  const tag = customNpmTag || npmTag;
  return {
    tag,
    version,
  };
}

export async function release(options: ReleaseNS.IOptions) {
  if (options.dryRun) {
    logger.info(`[dry-run] enabled`);
  }

  const cwd = options.cwd ?? process.cwd();

  const config = resolveLernaConfig(cwd);

  if (!config?.path) {
    throw new Error('"mono.json" or "lerna.json" doesn\'t exist');
  }

  const lernaConfig: ReleaseNS.ILernaConfig = config.data;

  if (lernaConfig.version === "independent") {
    throw new Error('"release" cannot be executed under "independent" mode');
  }

  if (typeof options.ready === "function") {
    await options.ready({ cwd });
  }

  logger.info(`${chalk.gray("Current version: ")}${lernaConfig.version}`);

  const { version, tag } = await selectVersionAndTag(lernaConfig.version);

  const { yes } = await inquirer.prompt([
    {
      name: "yes",
      message: `Confirm releasing ${version} (${tag})?`,
      type: "list",
      choices: ["N", "Y"],
    },
  ]);

  if (yes === "N") {
    logger.info("cancelled.");
    return;
  }

  /**
   * Set this env for subsequent build process.
   */
  process.env.MONO_RELEASE_VERSION = version;

  /**
   * Execute custom build script before release.
   */
  if (options.build) {
    const buildScript =
      typeof options.build === "string" ? options.build : "npm run build";
    if (options.dryRun) {
      logger.info("[dry-run] run build with: " + chalk.gray(buildScript));
    } else {
      const [command, ...args] = buildScript.split(" ");
      await execa(command, args, {
        shell: true,
        cwd: options.cwd,
        stdio: "inherit",
        env: process.env,
      });
    }
  }

  let releaseArguments: string[] = [
    "publish",
    version,
    "--exact",
    "--force-publish",
    "--dist-tag",
    tag,
  ];

  if (typeof options.modifyReleaseArguments === "function") {
    releaseArguments = options.modifyReleaseArguments({
      arguments: releaseArguments,
      version,
      tag,
    });
  }

  if (options.ignoreScripts) {
    releaseArguments.push("--ignore-scripts");
  }

  ensureLernaConfig(cwd);

  const lernaPath = require.resolve("lerna/cli.js", {
    paths: [cwd, __dirname],
  });
  if (options.dryRun) {
    logger.info("[dry-run] run publish with:");
    logger.info(
      "[dry-run] " + chalk.gray(lernaPath + " " + releaseArguments.join(" "))
    );
    logger.info("[dry-run] enter patch process");
    logger.info("[dry-run] finish patch process");
  } else {
    try {
      await execa(lernaPath, releaseArguments, {
        stdio: "inherit",
      });
    } catch (e) {
      console.log(e);

      await patch({
        cwd,
        version,
        tag,
        runInBand: options.runInBand,
        ignoreScripts: options.ignoreScripts,
      });
    }
  }

  if (typeof options.released === "function") {
    await options.released({ cwd, version, tag });
  }

  cleanLernaConfig(cwd);

  if (options.changelog) {
    const changelogOptions: ChangelogNS.IOptions = {
      cwd,
      beautify: true,
      commit: true,
      gitPush: true,
      attachAuthor: true,
      authorNameType: "email",
    };
    if (options.dryRun) {
      logger.info(
        "[dry-run] generate change with " +
          chalk.gray(JSON.stringify(changelogOptions, null, 2))
      );
    } else {
      await changelog();
    }
  }
}
