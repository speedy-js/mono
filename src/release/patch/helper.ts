/**
 * Module dependencies
 */
import chalk from "chalk";
import boxen, { BorderStyle } from "boxen";
import textTable from "text-table";
import stringWidth from "string-width";
import { NpmNS } from "../types";

/**
 * Get visible release status log
 */
export function getVisibleReleaseStatusLog(
  pkgs: NpmNS.IRemotePackageInfo[],
  currentVersion: string,
  currentTag: string
) {
  const table = pkgs.map((pkg) => {
    if (pkg.version === currentVersion) {
      return [pkg.name, pkg.version, "Yes", "-"].map((v) => chalk.dim(v));
    }

    return [
      chalk.bold(chalk.red(pkg.name)),
      chalk.bold(chalk.red(pkg.version)),
      chalk.bold(chalk.red("NO")),
      chalk.bold(chalk.red(currentVersion)),
    ];
  });

  table.unshift(
    [
      "Package Name",
      `Remote Version (tag: ${currentTag})`,
      "Published",
      "Expected Version",
    ].map((v) => chalk.dim(v))
  );

  return boxen(
    textTable(table, {
      stringLength: stringWidth,
    }),
    {
      borderStyle: BorderStyle.Round,
      padding: 1,
      dimBorder: true,
    }
  )
}
