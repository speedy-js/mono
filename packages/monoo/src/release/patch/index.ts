/**
 * Module dependencies
 */
import { join } from "path";
import { readFileSync, writeFileSync } from "fs";
import inquirer from "inquirer";
import { fetchPackageVersion, logger } from "../shared";
import { PatchNS } from "../types";
import { getVisibleReleaseStatusLog } from "./helper";
import { IMonorepoPackageWithRemoteInfo } from "@monoo/types";
import {
  chalk,
  execa,
  resolveLernaConfig,
  loadMonorepoPackages,
} from "@monoo/shared";

/**
 * Remove duplicate gitHead field when "lerna publish" fails.
 *
 * @param {String} path
 * @returns {void}
 */

export function removeGitHead(path: string) {
  const GIT_HEAD_REG = /,[\n\t\s]*"gitHead":\s"[^"]+"/;
  let content = readFileSync(path, "utf-8");
  content = content.replace(GIT_HEAD_REG, "");
  writeFileSync(path, content, "utf-8");
}

/**
 * Enter a interactive cli to know the current situation of `lerna publish`.
 *
 * @param {String} cwd
 * @param {String} version
 * @param {String} tag
 * @returns {Promise<void>}
 */

export async function patch(options: PatchNS.IOptions): Promise<void | never> {
  const { cwd, tag, runInBand = false, ignoreScripts = false } = options;

  let { version } = options;

  logger.info("Patch Started");

  if (!tag) {
    throw new Error('[MONOO] "tag" is required for "patch"');
  }

  const lernaConfig = resolveLernaConfig(cwd).data;

  if (!version) {
    try {
      version = lernaConfig?.version;
    } catch (e) {
      // dot not handle it for now
    }
  }

  logger.info(`Version: ${chalk.cyan(version)}`);
  logger.info(`Tag: ${chalk.cyan(tag)}`);

  const pkgs = await loadMonorepoPackages(cwd, lernaConfig);
  pkgs.forEach((pkg) => removeGitHead(join(pkg.dir, "package.json")));

  if (pkgs.every((pkg) => pkg.version === version)) {
    return console.log(
      `${chalk.cyan("‚ùØ ")}${chalk.gray(
        "[MONOO]"
      )} Do not need "patch" since all packages've been published correctly! `
    );
  }

  const remotePkgs: IMonorepoPackageWithRemoteInfo[] = await Promise.all(
    pkgs.map<Promise<IMonorepoPackageWithRemoteInfo>>(async (pkg) => {
      return {
        ...pkg,
        remoteVersion: await fetchPackageVersion(pkg.name, tag),
      };
    })
  );

  const visibleReleaseStatusLog = getVisibleReleaseStatusLog(
    remotePkgs,
    version,
    tag
  );

  console.log(visibleReleaseStatusLog);

  const { yes } = await inquirer.prompt([
    {
      name: "yes",
      message: "Continue to patch",
      type: "list",
      choices: ["N", "Y"],
    },
  ]);

  if (yes === "Y") {
    const patchedPkgs = remotePkgs.filter(
      (pkg) => pkg.remoteVersion !== version
    );

    const releasePkg = async (pkg: IMonorepoPackageWithRemoteInfo) => {
      const commandCwd = join(cwd, `packages/${pkg.dir}`);
      console.log(chalk.gray(`$ npm publish --tag ${tag} # ${commandCwd}`));
      const npmArgs = ["publish", "--tag", tag];
      if (ignoreScripts) {
        npmArgs.push("--ignore-scripts");
      }
      await execa("npm", npmArgs, {
        // silent: true,
        stdio: "inherit",
        cwd: commandCwd,
      });
      console.log(`+ ${pkg.name}@${version}\n`);
    };

    if (runInBand) {
      for (const pkg of patchedPkgs) {
        await releasePkg(pkg);
      }
    } else {
      await Promise.all(patchedPkgs.map(releasePkg));
    }
  }
}
