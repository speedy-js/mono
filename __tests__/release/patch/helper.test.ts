/**
 * Module dependencies
 */
import { getVisibleReleaseStatusLog } from "../../../src/release/patch/helper";
import { NpmNS } from "../../../src/release/types";

it("getVisibleReleaseStatusLog", () => {
  const log = getVisibleReleaseStatusLog(
    [
      {
        name: "@scope/foo",
        version: "2.1.2",
      },
      {
        name: "@scope/bar",
        version: "2.1.1",
      },
      {
        name: "@scope/baz",
        version: "2.1.2",
      },
      {
        name: "@scope/qux",
        version: "2.1.1",
      },
    ] as NpmNS.IRemotePackageInfo[],
    "2.1.2",
    "next"
  );
  console.log(log);
  expect(require("strip-ansi")(log)).toMatchSnapshot();
});
