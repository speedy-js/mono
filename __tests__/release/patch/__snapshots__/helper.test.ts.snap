// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getVisibleReleaseStatusLog 1`] = `
"╭───────────────────────────────────────────────────────────────────────────╮
│                                                                           │
│   Package Name  Remote Version (tag: next)  Published  Expected Version   │
│   @scope/foo    2.1.2                       Yes        -                  │
│   @scope/bar    2.1.1                       NO         2.1.2              │
│   @scope/baz    2.1.2                       Yes        -                  │
│   @scope/qux    2.1.1                       NO         2.1.2              │
│                                                                           │
╰───────────────────────────────────────────────────────────────────────────╯"
`;
